### PR TITLE
Move Nessie Client construction into a JUnit5 extension

### PR DESCRIPTION
**Pull request by dimas-b**
This is to unify Nessie Client handling in preparation for API v2 tests.

In general, this isolates the construction of the client, which is largely dependent on the test env. from the test code itself, which operates at the Nessie java client / API level.

NessieApiProvider is needed only to work around some @QuarkusTest oddities as described in source comments in NessieClientResolver.

### CODE REVIEW

File: servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieJaxRsExtension.java [hunk 1]
```diff
@@ -159,6 +167,17 @@ public Object resolveParameter(
             + parameterContext.getTarget());
   }
 
+  @Override
+  protected URI findBaseUri(ExtensionContext extensionContext) {
+    EnvHolder env = extensionContext.getStore(NAMESPACE).get(EnvHolder.class, EnvHolder.class);
+    if (env == null) {
+      throw new ParameterResolutionException(
+          "Nessie JaxRs env. is not initialized in " + extensionContext.getUniqueId());
+    }
```
Comment (XN137) [inline servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieJaxRsExtension.java:176]: this logic seems duplicated, instead we should probably have a `private EnvHolder getEnv()` helper method that includes this check and gets used everwhere, wdyt ?

File: clients/client-testextension/src/main/java/org/projectnessie/client/ext/NessieClientResolver.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    if (isNessieClient(parameterContext)) {
+      return makeClient(extensionContext);
+    }
+
+    throw new IllegalStateException("Unsupported parameter: " + parameterContext);
+  }
+
+  private NessieApiProvider makeClient(ExtensionContext extensionContext) {
```
Comment (XN137) [inline clients/client-testextension/src/main/java/org/projectnessie/client/ext/NessieClientResolver.java:69]: nit: naming wise this also reads odd... makeClient returns something that can build an api instance :sweat_smile: 
maybe `NessieApiFactory` and `get` becomes `build(..)` ? "provider" sounds like you cant influence the returned object imo

**Review by XN137 — APPROVED** [review]
some suggestions for internals and naming but overall seems like a good step in the right direction, so +1

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRest.java [hunk 3]
```diff
@@ -50,17 +50,13 @@ public abstract class AbstractRest {
...
-  protected void initApi(URI nessieApiUri) {
-    NessieApiV1 api = HttpClientBuilder.builder().withUri(nessieApiUri).build(NessieApiV1.class);
-    initApi(api);
-  }
-
-  protected void initApi(NessieApiV1 api) {
-    this.api = api;
+  @BeforeEach
+  void initApi(NessieApiProvider api) {
+    this.api = Objects.requireNonNull(api.get());
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRest.java:55]: will add `@Nonnull`

File: servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieUri.java [hunk 4]
```diff
@@ -24,8 +24,11 @@
 /**
  * Annotation for JUnit5 method parameters that need a URI to the Nessie server managed by {@link
  * NessieJaxRsExtension}.
+ *
+ * @deprecated Use {@link org.projectnessie.client.ext.NessieUri} instead.
  */
 @Target({ElementType.FIELD, ElementType.PARAMETER})
 @Retention(RetentionPolicy.RUNTIME)
 @Inherited
+@Deprecated
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieUri.java:33]: Integration tests (from Iceberg) still use it, and NesQEIT runs latest Iceberg against latest Nessie.

File: servers/quarkus-server/src/test/java/org/projectnessie/server/QuarkusNessieUriResolver.java [hunk 5]
```diff
@@ -18,40 +18,18 @@
...
-
   @Override
-  public Object resolveParameter(ParameterContext paramCtx, ExtensionContext extensionCtx)
-      throws ParameterResolutionException {
-    if (isQuarkusNessieUriParam(paramCtx)) {
-      return URI.create(getNessieApiV1Url());
-    }
-    throw new ParameterResolutionException(
-        "Unsupported parameter " + paramCtx.getParameter() + " on " + paramCtx.getTarget());
+  protected URI findBaseUri(ExtensionContext extensionContext) {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply servers/quarkus-server/src/test/java/org/projectnessie/server/QuarkusNessieUriResolver.java:32]: Renaming for `getBaseUri`

File: servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRest.java [hunk 6]
```diff
@@ -50,17 +50,13 @@ public abstract class AbstractRest {
...
   }
 
   public NessieApiV1 getApi() {
-    return Objects.requireNonNull(api, "Tests need to call initApi in @BeforeEach");
+    return Objects.requireNonNull(api, "Nessie API field was not injected.");
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRest.java:59]: Removing extra checks :+1:

File: clients/client-testextension/src/main/java/org/projectnessie/client/ext/NessieClientResolver.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+ * A base class for extensions that manage a Nessie test execution environment. This class injects
+ * Nessie java client instances and/or URIs based on the specific environment details provided by
+ * concrete subclasses.
+ *
+ * @see NessieUri
+ * @see NessieApiProvider
+ */
+public abstract class NessieClientResolver implements ParameterResolver {
+
+  protected abstract URI findBaseUri(ExtensionContext extensionContext);
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply clients/client-testextension/src/main/java/org/projectnessie/client/ext/NessieClientResolver.java:37]: Renaming to `getBaseUri`... `Nessie` is implied from the class name and it's not likely to clash with some other "base URI" :)

File: clients/iceberg-views/src/test/java/org/apache/iceberg/nessie/BaseIcebergTest.java [hunk 8]
```diff
@@ -94,9 +94,9 @@ private void resetData() throws NessieConflictException, NessieNotFoundException
   }
 
   @BeforeEach
-  public void beforeEach(@NessieUri URI x) throws IOException {
-    uri = x.toString();
-    this.api = HttpClientBuilder.builder().withUri(uri).build(NessieApiV1.class);
+  public void beforeEach(NessieApiProvider apiProvider, @NessieUri URI uri) throws IOException {
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply clients/iceberg-views/src/test/java/org/apache/iceberg/nessie/BaseIcebergTest.java:97]: I'd like to keep `NessieApiProvider` free from dependencies on a specific client protocol. Granted some tests assume it's HTTP, but at least it is not forced by the parameter type.

File: clients/client-testextension/src/main/java/org/projectnessie/client/ext/NessieClientResolver.java (hunk 2 above)
Comment (dimas-b) [reply clients/client-testextension/src/main/java/org/projectnessie/client/ext/NessieClientResolver.java:69]: renaming...

File: clients/client-testextension/src/main/java/org/projectnessie/client/ext/NessieApiProvider.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+package org.projectnessie.client.ext;
+
+import org.projectnessie.client.api.NessieApiV1;
+
+/**
+ * Interface for declaring parameters in test methods wishing to operate on a Nessie Client
+ * pre-configured to talk to the Nessie Server running in the current test environment.
+ *
+ * <p>An implementation of this interface will be injected into test method parameters by {@link
+ * NessieClientResolver}.
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply clients/client-testextension/src/main/java/org/projectnessie/client/ext/NessieApiProvider.java:25]: renaming...

File: servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieJaxRsExtension.java [hunk 10]
```diff
@@ -125,25 +127,36 @@ public void beforeEach(ExtensionContext extensionContext) {
...
+        || parameterContext.isAnnotated(NessieUri.class)
         || parameterContext.isAnnotated(NessieSecurityContext.class)
         || parameterContext.isAnnotated(NessieAccessChecker.class);
   }
 
-  @Override
-  public Object resolveParameter(
-      ParameterContext parameterContext, ExtensionContext extensionContext)
-      throws ParameterResolutionException {
+  private EnvHolder getEnv(ExtensionContext extensionContext) {
```
Comment (XN137) [inline servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieJaxRsExtension.java:136]: i think this helper can be used in 3 more places

**Review by XN137 — DISMISSED** [review]
thanks for the update, still lgtm

File: servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieJaxRsExtension.java (hunk 10 above)
Comment (dimas-b) [reply servers/jax-rs-testextension/src/main/java/org/projectnessie/jaxrs/ext/NessieJaxRsExtension.java]: ah, missed that - updated

**Review by XN137 — APPROVED** [review]
still +1 after 2rd commit

**Pull request merged**